### ISSUE TITLE
Lint request: flag useless assignments

### ISSUE THREAD

**Issue by kentcb** [dart-lang/linter#1597]
**Describe the rule you'd like to see implemented**

Just spent 10 minutes debugging a problem that got introduced during a refactor:

```dart
  void _showIfHidden(Rect globalBounds) {
    if (!show) {
      setState(() {
          globalBounds = globalBounds;
          show = true;
        });
    }
  }
```

The `globalBounds = globalBounds` is a no-op. It's supposed to be `this.globalBounds = globalBounds`.

It would be nice to flag these useless assignments because they're an easy mistake to make, are never intentional, and can waste a bunch of developer time.

**Examples**

See above.

**Comment by pq** [comment]
I'm a little surprised we don't do this already in analyzer...  (cc @bwilkerson for thoughts)

**Comment by pq** [comment]
(Needless to say, thanks for the suggestion @kentcb!)

**Comment by bwilkerson** [comment]
I'm also a bit surprised. We can make it a hint if we want. Either way we should limit it to assignments involving local variables (which includes parameters).

**Comment by pq** [comment]
@bwilkerson: if you can't think of a reason not to, a hint might be preferable.  Unless we can think of a reason someone may not want to opt-in?

**Comment by bwilkerson** [comment]
I can't think of any reason not to make it a hint. As was pointed out by the OP, it's never intentional.

**Comment by pq** [comment]
Moved to https://github.com/dart-lang/sdk/issues/37162.

**Comment by oprypin** [comment]
This came up internally as a lint that could've prevented a bad change that caused problems in production.

This issue was moved to main repo as something that's "never intentional".

But there *can* be a valid reason to assign to itself - invoke some update logic on a setter.

```dart
class C {
  int _x = 5;
  
  int get x => _x;

  set x(int x) {
    _x = x;
    _update();
  }
  
  void _update() {
    print('ok');
  }

  void test() {
    x = x;
  }
}

void main() {
  C().test();
}
```

**Comment by srawlins** [comment]
I still think this should be Moved to https://github.com/dart-lang/sdk/issues/37162.

**Comment by oprypin** [comment]
Now I actually see the distinction - maybe a subset of this **should** be an analyzer check. For the kinds of variables that can never be backed by a setter (maybe that's just local variables), there can indeed never be false positives. I was just thinking that the broader check is still useful and so is a good lint.

To me the easiest outcome is just to land the lint in #4067, but maybe this is something to think about first.

**Comment by bwilkerson** [comment]
I'm undecided. The local variable case has zero false positives, so it seems reasonable to enable it by default, but it might be easier for external users if there's one diagnostic about this kind of issue and not have to learn that half of it is enabled by default and half of it needs to be explicitly enabled.

**Comment by pq** [comment]
> it might be easier for external users if there's one diagnostic about this kind of issue

👍 

This is my gut instinct.
